Telecommands to Jump to Golden Copy and Switch Active Flash Bank

NuclearTea opened the pull request:
#24

DeflateAwning commented:
Please ensure pull requests are opened with a title which describes the change they implement.

NuclearTea commented:
I will add instructions on how to test in a bit.

NuclearTea commented:
@DeflateAwning After discussing with you, you recommended that I look for libraries that can assist with bootloader things. After searching for a while I found:

https://github.com/STMicroelectronics/stm32-mw-openbl

After looking into it, and playing around with out, I found out that for the most part, it does not give us anything really. It has some handy functions to read and write to internal flash memory as well as a jump to application function, however, most of the functions that are required must be implemented ourselves. It merely gives us templates and a couple of helper functions. If we really wanted, I could copy over the files that are for reading and writing internal flash memory however, the rest is really not useful for us.

Additionally, I have already compared my own implementation and their implementation of these functions and they are, for the most part, identical. Actually, I have much more error handling than they do in every function so that is actually and advantage for our source code. There is 1 place where I will update our implementation (which is to do with writing to the internal flash memory and when the data itself is not 8 bytes long) but other than that, its nearly identical.

For some reference: 
| Feature                     | Our Implementation (File)     | Their Implementation (File)                                                                                                                                                                                                 |
|-----------------------------|-------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| JumpToApplication           | Bootloader.c                  | Core/Interfaces/Patterns/Flash/flash_interface.c (https://github.com/STMicroelectronics/stm32-mw-openbl/blob/7689c2de09d3e107923bb2c5971eb09d9299a76e/Interfaces/Patterns/FLASH/flash_interface.c#L170)                     |
| Deinit Peripherals          | Bootloader.c                  | Core/openbl_core.c (https://github.com/STMicroelectronics/stm32-mw-openbl/blob/7689c2de09d3e107923bb2c5971eb09d9299a76e/Core/openbl_core.c#L57) I can’t even find the implementation for the OpenBootloader_DeInit function |
| Flash Lock/Unlock/OB_Unlock | stm32_internal_flash_driver.c | Core/Interfaces/Patterns/Flash/flash_interface.c (https://github.com/STMicroelectronics/stm32-mw-openbl/blob/7689c2de09d3e107923bb2c5971eb09d9299a76e/Interfaces/Patterns/FLASH/flash_interface.c#L79)                      |
| Flash Read/Write            | stm32_internal_flash_driver.c | Core/Interfaces/Patterns/Flash/flash_interface.c (https://github.com/STMicroelectronics/stm32-mw-openbl/blob/7689c2de09d3e107923bb2c5971eb09d9299a76e/Interfaces/Patterns/FLASH/flash_interface.c#L109)                     |



Would love your input on this!

NuclearTea commented:
## Notes
1. You will need the stm32-for-vscode extension (up-to-date version). **It is important that it is up to date**
2. You will need the `STM32CubeProgrammer` which is downloadable from the stm32 website

# Testing Steps
1. Checkout this branch `ali-restore-golden-copy-i24`
2. Open a serial terminal with the appropriate UART configurations.
3. `Clean Build` the project and `Flash` it to your STM32
4. Ensure that it is working as normal.
5. Open the STM32CubeProgrammer and connect it to you STM32 Microcontroller
	1. ![image](https://github.com/user-attachments/assets/34868303-2f37-41ab-b212-9ccff0b6a16c)
	2. This will cause your STM to pause. Your serial test will not be receiving anything new. (You can reset your STM32 manually by pressing the reset button on the microcontroller to resume it working while still being connected).
	3. Click the `Upload` button on the left toolbar. It is the second from the top.
		1. ![image](https://github.com/user-attachments/assets/f627d18b-9704-4f56-bd9f-bb3baa241252)
	4. This will show the following screen:
	   ![image](https://github.com/user-attachments/assets/b38f06e8-d450-453c-abff-501f71a9e964)
		1. Underneath the `red` 1. drawn, you can select the file path to the firmware .bin file. I would recommend making the following modifications to the firmware.
			1. Open `rtos_tasks.c` and modify the `TASK_DEBUG_print_heartbeat` function.
				1. In the while loop at the bottom: 
				2. ![image](https://github.com/user-attachments/assets/54ba022b-f24d-44e4-a8f5-df54bd1387a8)
				3. Edit line 68 in the screenshot above so that it says: `FrontierSat version 2 time: ....`. We are only adding the words: `version 2` infront of `FrontierSat` and not changing anything else.
		2. Click the `Build` button.
		3. **DO NOT FLASH**
		4. As shown in the above CubeProgrammer screenshot, underneath the `red` 1., please enter the path to this .bin file. It should be in the build/debug folder of the firmware project.
		5. To the left of the `red` 2., ensure the value is: `0x8100000`.
		6. In the checkboxes below the `red` 2., please ensure **Only** **the**`Verify programming` **is selected**.
		7. Click the `Start Program` button
		8. Most likely after completing, your stm will be paused. Press the reset button on your STM32 microcontroller.
		9. At this point, it should just like before, running the exact same as before. If you followed my recommendation before, you can confirm this as the heartbeat should be saying `FrontierSat time: ...`
		10. We can now jump to the new program by sending the following telecommand: `CTS1+stm32_internal_flash_set_dual_bank_boot(1)!`
			1. This enables `dual_bank_mode` which changes the start address to Flash Bank 2.
			2. Following is an example:
![image](https://github.com/user-attachments/assets/9130cb5f-faf6-4ea3-a081-29e550227073)
		11. To go back, we can just send the same telecommand but with 0 as the param: `CTS1+stm32_internal_flash_set_dual_bank_boot(0)!`

**PDF Version:** [Testing Steps for Jumping to Golden Copy.pdf](https://github.com/user-attachments/files/18877245/Testing.Steps.for.Jumping.to.Golden.Copy.pdf)

NuclearTea commented:
@DeflateAwning I realized that the entire bootloader section is technically not necessary actually though it may become useful later. Do you want me to remove it or keep it and see if #25 will require it? From what I discovered, it does not seem like it will be needed.

NuclearTea commented:
Removed all bootloader stuff, sqashed commits to make it concise

DeflateAwning commented on firmware/Core/Src/telecommands/stm32_internal_flash_telecommand_defs.c:
```diff
@@ -114,4 +114,78 @@ uint8_t TCMDEXEC_stm32_internal_flash_erase(const char *args_str, TCMD_Telecomma
+/// - Arg 0: A 1 or 0. 1 enables dual bank boot, 0 disables
+/// @param response_output_buf Prints error if it occurs
+/// @return 0 on success, > 0 otherwise
+uint8_t TCMDEXEC_stm32_internal_flash_set_dual_bank_boot(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel, char *response_output_buf, uint16_t response_output_buf_len)
```
What's the default? Why would anyone want to run this telecommand? 

I'm sure there's a reason, but it should go in this docstring.

DeflateAwning commented on firmware/Core/Src/telecommands/stm32_internal_flash_telecommand_defs.c:
```diff
@@ -114,4 +114,78 @@ uint8_t TCMDEXEC_stm32_internal_flash_erase(const char *args_str, TCMD_Telecomma
+
+    snprintf(response_output_buf, response_output_buf_len,
+             "{\n"
+             "  \"OptionType\": %lu,\n"
```
Generally we don't put the trailing newline whitespaces and instead rely on the JSON autoformatter in the Ground Support Terminal. Up to you though.

DeflateAwning commented on firmware/Core/Src/system_stm32l4xx.c:
```diff
@@ -123,7 +123,7 @@
      anywhere in Flash or Sram, else the vector table is kept at the automatic
      remap of boot address selected */
-/* #define USER_VECT_TAB_ADDRESS */
+#define USER_VECT_TAB_ADDRESS 
```
Can you please leave a comment like:
```
// CTS-SAT-1: The `USER_VECT_TAB_ADDRESS` macro is defined because...
```

DeflateAwning commented on firmware/Core/Src/stm32_internal_flash_drivers/stm32_internal_flash_drivers.c:
```diff
@@ -138,4 +139,80 @@ uint8_t STM32_internal_flash_erase(uint16_t start_page_erase, uint16_t number_of
+}
+
+/// @brief Sets the dual bank mode on or off depending on param
+/// @param dual_bank_mode 1 or 0
```
Duplicate the comment from the telecommand docstring please

DeflateAwning reviewed (changes requested):
All seems fine. Does this PR add any features to CTS-SAT-1, or is it all pre-work for upcoming over-the-air upgrades (if implemented)

NuclearTea commented:
# Updated Testing Instructions
## Notes
1. You will need the stm32-for-vscode extension (up-to-date version). **It is important that it is up to date**
2. You will need the `STM32CubeProgrammer` which is downloadable from the stm32 website

# Testing Steps
1. Checkout this branch `ali-restore-golden-copy-i24`
2. Open a serial terminal with the appropriate UART configurations.
3. `Clean Build` the project and `Flash` it to your STM32
4. Ensure that it is working as normal.
5. Open the STM32CubeProgrammer and connect it to you STM32 Microcontroller
	1. ![image](https://github.com/user-attachments/assets/34868303-2f37-41ab-b212-9ccff0b6a16c)
	2. This will cause your STM to pause. Your serial test will not be receiving anything new. (You can reset your STM32 manually by pressing the reset button on the microcontroller to resume it working while still being connected).
	3. Click the `Upload` button on the left toolbar. It is the second from the top.
		1. ![image](https://github.com/user-attachments/assets/f627d18b-9704-4f56-bd9f-bb3baa241252)
	4. This will show the following screen:
	   ![image](https://github.com/user-attachments/assets/b38f06e8-d450-453c-abff-501f71a9e964)
		1. Underneath the `red` 1. drawn, you can select the file path to the firmware .bin file. I would recommend making the following modifications to the firmware.
			1. Open `rtos_tasks.c` and modify the `TASK_DEBUG_print_heartbeat` function.
				1. In the while loop at the bottom: 
				2. ![image](https://github.com/user-attachments/assets/54ba022b-f24d-44e4-a8f5-df54bd1387a8)
				3. Edit line 68 in the screenshot above so that it says: `FrontierSat version 2 time: ....`. We are only adding the words: `version 2` infront of `FrontierSat` and not changing anything else.
		2. Click the `Build` button.
		3. **DO NOT FLASH**
		4. As shown in the above CubeProgrammer screenshot, underneath the `red` 1., please enter the path to this .bin file. It should be in the build/debug folder of the firmware project.
		5. To the left of the `red` 2., ensure the value is: `0x8100000`.
		6. In the checkboxes below the `red` 2., please ensure **Only** **the**`Verify programming` **is selected**.
		7. Click the `Start Program` button
		8. Most likely after completing, your stm will be paused. Press the reset button on your STM32 microcontroller.
		9. At this point, it should just like before, running the exact same as before. If you followed my recommendation before, you can confirm this as the heartbeat should be saying `FrontierSat time: ...`
		10. We can now jump to the new program by sending the following telecommand: `CTS1+stm32_internal_flash_set_dual_bank_boot(2)!`
			1. This enables `dual_bank_mode` which changes the start address to Flash Bank 2.
			2. Following is an example:
![image](https://github.com/user-attachments/assets/56e67ac2-3aef-4305-9ec0-bfa1953745d9)
		11. To go back, we can just send the same telecommand but with 0 as the param: `CTS1+stm32_internal_flash_set_dual_bank_boot(1)!`

**PDF Version:**
[Testing Steps for Jumping to Golden Copy.pdf](https://github.com/user-attachments/files/19239006/Testing.Steps.for.Jumping.to.Golden.Copy.pdf)
)

NuclearTea commented:
@DeflateAwning Addressed all PR comments. Regarding your question about whether this adds a new feature or is pre-work. Techincally, if we had a "golden copy" flashed on the OBC, then this would be a feature as you could boot from it. It does not contain anything to do with loading a file from the memory module and storing it in our flash memory as that is: #18, #25

DeflateAwning reviewed (approved):
All seems reasonable to me! Didn't test it though.

I'd recommend getting @robotoshi to check this next time he's around to independently verify that it works (on an OBC, probably), and also to validate that the documentation/guide is reasonable.

kev9268 commented:
@DeflateAwning 

After setting up the test prerequisites I inputted CTS1+stm32_internal_flash_set_active_flash_bank(2)!

Output of CTS1+stm32_internal_flash_set_active_flash_bank(2)! :
```
CTS1+stm32_internal_flash_set_active_flash_bank(2)!
0000000000000+0000022083_N [TELECOMMAND:DEBUG]: Parsed telecommand (len=52): 'CTS1+stm32_internal_flash_set_active_flash_bank(2)!
'
0000000000000+0000022234_N [TELECOMMAND:NORMAL]: Executing telecommand from agenda slot 0, sent at tssent=0, scheduled for tsexec=0.
0000000000000+0000022246_N [TELECOMMAND:NORMAL]: ðŸš€ Executing telecommand 'stm32_internal_flash_set_active_flash_bank'.


MX_Init() done
0000000000000+0000000066_N [EPS:ERROR]: EPS UART Reception Error: 4
0000000000000+0000000170_N [OBC:NORMAL]: TASK_DEBUG_print_heartbeat() -> started (booted).
0000000000000+0000000178_N [OBC:NORMAL]: Reset reason: EXTERNAL_RESET_PIN.
0000000000000+0000005170_N [EPS:NORMAL]: Setting OBC time based on EPS time because last_source == TIM_SOURCE_NONE
0000000000000+0000005233_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000005240_N [EPS:ERROR]: EPS_set_obc_time_based_on_eps_time() -> Error
0000000000000+0000010222_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000010229_N [EPS:ERROR]: EPS_CMD_watchdog() -> Error: 3
0000000000000+0000010248_N [EPS:NORMAL]: Setting OBC time based on EPS time because last_source == TIM_SOURCE_NONE
0000000000000+0000010311_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000010318_N [EPS:ERROR]: EPS_set_obc_time_based_on_eps_time() -> Error
0000000000000+0000012155_N [OBC:DEBUG]: Heartbeat: FrontierSat v2 time: 19700101T00:00:12.155, Uptime: 12155 ms
```

From the output we can see the heartbeat prints out FrontierSat v2 as expected

Now switching back by inputting the tcmd  CTS1+stm32_internal_flash_set_active_flash_bank(1)!

Output of CTS1+stm32_internal_flash_set_active_flash_bank(1)! 
```
0000000000000+0000254169_N [OBC:DEBUG]: Heartbeat: FrontierSat v2 time: 19700101T00:04:14.169, Uptime: 254169 ms

0000000000000+0000259070_N [EPS:NORMAL]: Setting OBC time based on EPS time because last_source == TIM_SOURCE_NONE
0000000000000+0000259133_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000259140_N [EPS:ERROR]: EPS_set_obc_time_based_on_eps_time() -> Error
CTS1+stm32_internal_flash_set_active_flash_bank(1)!
0000000000000+0000259828_N [TELECOMMAND:DEBUG]: Parsed telecommand (len=52): 'CTS1+stm32_internal_flash_set_active_flash_bank(1)!
'
0000000000000+0000259899_N [TELECOMMAND:NORMAL]: Executing telecommand from agenda slot 0, sent at tssent=0, scheduled for tsexec=0.
0000000000000+0000259912_N [TELECOMMAND:NORMAL]: ðŸš€ Executing telecommand 'stm32_internal_flash_set_active_flash_bank'.


MX_Init() done
0000000000000+0000000066_N [EPS:ERROR]: EPS UART Reception Error: 4
0000000000000+0000000170_N [OBC:NORMAL]: TASK_DEBUG_print_heartbeat() -> started (booted).
0000000000000+0000000178_N [OBC:NORMAL]: Reset reason: EXTERNAL_RESET_PIN.
0000000000000+0000005170_N [EPS:NORMAL]: Setting OBC time based on EPS time because last_source == TIM_SOURCE_NONE
0000000000000+0000005233_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000005240_N [EPS:ERROR]: EPS_set_obc_time_based_on_eps_time() -> Error
0000000000000+0000010222_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000010229_N [EPS:ERROR]: EPS_CMD_watchdog() -> Error: 3
0000000000000+0000010248_N [EPS:NORMAL]: Setting OBC time based on EPS time because last_source == TIM_SOURCE_NONE
0000000000000+0000010311_N [EPS:ERROR]: EPS->OBC: timeout before first byte received
0000000000000+0000010318_N [EPS:ERROR]: EPS_set_obc_time_based_on_eps_time() -> Error
0000000000000+0000012155_N [OBC:DEBUG]: Heartbeat: FrontierSat time: 19700101T00:00:12.155, Uptime: 12155 ms
```

From the output we see that the heartbeat print has returned back to only FrontierSat as expected in the testing instructions.
